File: Chapter 1/code/dining-philosophers.cpp
#include<bits/stdc++.h>
#include<semaphore.h> // POSIX semaphore

int main() {
    const int n = 5;  // number of philosophers
    const int m = 10; // number of times each philosopher eats
    // make semaphores
    sem_t* sticks = new sem_t[n];
    for (int i = 0; i < n; i++) {
        sem_init(&sticks[i], 0, 1);
    }
    std::vector<std::thread> trds;
    // making each philosopher hungry m times
    for (int i = 0; i < n; i++) {
        // create, start and add a new thread
        trds.push_back(std::thread([m, i, sticks]() {
            int idx1 = i+1, idx2 = i;
            if (i == n-1) idx1 = i, idx2 = 0;
            auto &stick1 = sticks[idx1], &stick2 = sticks[idx2];
            for (int M = m; M--; ) {
                sem_wait(&stick1);
                printf("pickup stick1 | done by: %d\n", i);
                sem_wait(&stick2);
                printf("pickup stick2 | done by: %d\n", i);
                printf("Philosopher %d is eating\n", i);
                usleep(1);
                sem_post(&stick1);
                usleep(1);
                sem_post(&stick2);
            }
        }));
    }
    // join threads
    for (auto& trd: trds) trd.join();
}

File: Chapter 1/code/dining-philosophers.py
import time
from multiprocessing import Process, Semaphore

n = 5  # number of philosophers
m = 10 # number of times each philosopher eats

# make semaphores
sticks = [Semaphore(1) for _ in range(n)]
procs = []

# making each philosopher hungry m times
for i in range(n):
    def f(i):
        stick1, stick2 = (sticks[i],sticks[0]) if i == n-1 else (sticks[i+1], sticks[i])
        for _ in range(m):
            stick1.acquire()
            print("pickup stick1 | done by:", i)
            stick2.acquire()
            print("pickup stick2 | done by:", i)
            print("Philosopher", i, "is eating")
            stick1.release()
            time.sleep(0.005)
            stick2.release()
    # create and start a new thread
    procs.append(Process(target = f, args = [i], daemon=True))
    procs[-1].start()

for proc in procs: proc.join()
